// element-react/dist/npm/es6/src/checkbox/CheckBoxButton.js
function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

import React from 'react';
import CheckBox from './CheckBox';

var CheckboxButton = function (_CheckBox) {
  _inherits(CheckboxButton, _CheckBox);

  function CheckboxButton() {
    _classCallCheck(this, CheckboxButton);

    return _possibleConstructorReturn(this, _CheckBox.apply(this, arguments));
  }

  CheckboxButton.prototype.render = function render() {
    var group = this.context.ElCheckboxGroup;

    return React.createElement(
      'label',
      { style: this.style(), className: this.className('el-checkbox-button', group.props.size ? 'el-checkbox-button--' + group.props.size : '', {
          'is-disabled': this.props.disabled,
          'is-checked': this.state.checked,
          'is-focus': this.state.focus
        }) },
      React.createElement('input', {
        className: 'el-checkbox-button__original',
        type: 'checkbox',
        checked: this.state.checked,
        disabled: this.props.disabled,
        onFocus: this.onFocus.bind(this),
        onBlur: this.onBlur.bind(this),
        onChange: this.onChange.bind(this)
      }),
      React.createElement(
        'span',
        { className: 'el-checkbox-button__inner', style: this.state.checked ? {
            boxShadow: '-1px 0 0 0 ' + group.props.fill,
            backgroundColor: group.props.fill || '',
            borderColor: group.props.fill || '',
            color: group.props.textColor || ''
          } : {} },
        this.state.label || this.props.children
      )
    );
  };

  return CheckboxButton;
}(CheckBox);

CheckboxButton.elementType = 'CheckboxButton';
export default CheckboxButton;
