// element-react/dist/npm/es6/libs/transition/index.js
function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

import React, { Component } from 'react';
import PropTypes from 'prop-types';
import ReactCSSTransitionGroup from 'react-transition-group/CSSTransitionGroup';

var Transition = function (_Component) {
  _inherits(Transition, _Component);

  function Transition() {
    _classCallCheck(this, Transition);

    return _possibleConstructorReturn(this, _Component.apply(this, arguments));
  }

  Transition.prototype.render = function render() {
    return React.createElement(ReactCSSTransitionGroup, {
      transitionName: this.props.name,
      transitionEnterTimeout: Number(this.props.duration),
      transitionLeaveTimeout: Number(this.props.duration),
      component: this.props.component,
      className: this.props.className,
      style: this.props.style
    }, this.props.children);
  };

  return Transition;
}(Component);

export default Transition;


Transition.propTypes = {
  name: PropTypes.string.isRequired,
  duration: PropTypes.oneOfType([PropTypes.number, PropTypes.string]),
  component: PropTypes.string,
  className: PropTypes.string,
  style: PropTypes.object
};

Transition.defaultProps = {
  duration: 300
};
