// element-react/dist/npm/es6/src/collapse/Collapse.js
function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

import React from 'react';
import { Component, PropTypes } from '../../libs';

var Collapse = function (_Component) {
  _inherits(Collapse, _Component);

  function Collapse(props) {
    _classCallCheck(this, Collapse);

    var _this = _possibleConstructorReturn(this, _Component.call(this, props));

    _this.state = {
      activeNames: [].concat(_this.props.value)
    };
    return _this;
  }

  Collapse.prototype.setActiveNames = function setActiveNames(activeNames) {
    var _this2 = this;

    activeNames = [].concat(activeNames);
    this.setState({ activeNames: activeNames }, function () {
      return _this2.props.onChange(activeNames);
    });
  };

  Collapse.prototype.handleItemClick = function handleItemClick(name) {
    var activeNames = this.state.activeNames;


    if (this.props.accordion) {
      this.setActiveNames(activeNames[0] && activeNames[0] === name ? '' : name);
    } else {
      if (activeNames.includes(name)) {
        this.setActiveNames(activeNames.filter(function (item) {
          return item !== name;
        }));
      } else {
        this.setActiveNames(activeNames.concat(name));
      }
    }
  };

  Collapse.prototype.render = function render() {
    var _this3 = this;

    var content = React.Children.map(this.props.children, function (child, idx) {
      var name = child.props.name || idx.toString();
      return React.cloneElement(child, {
        isActive: _this3.state.activeNames.includes(name),
        key: idx,
        name: name,
        onClick: function onClick(item) {
          return _this3.handleItemClick(item);
        }
      });
    });
    return React.createElement(
      'div',
      { className: 'el-collapse' },
      content
    );
  };

  return Collapse;
}(Component);

Collapse.defaultProps = {
  value: [],
  onChange: function onChange() {}
};
export default Collapse;


Collapse.propTypes = {
  accordion: PropTypes.bool,
  value: PropTypes.oneOfType([PropTypes.array, PropTypes.string]),
  onChange: PropTypes.func
};
