// element-react/dist/npm/es6/src/date-picker/TimeSelect.js
var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

import React from 'react';

import { PropTypes } from '../../libs';
import BasePicker from './BasePicker';

import TimeSelectPanel from './panel/TimeSelectPanel';

var TimeSelect = function (_BasePicker) {
  _inherits(TimeSelect, _BasePicker);

  _createClass(TimeSelect, null, [{
    key: 'propTypes',
    get: function get() {
      var result = Object.assign({}, {
        start: PropTypes.string,
        end: PropTypes.string,
        step: PropTypes.string,
        minTime: PropTypes.instanceOf(Date)
      }, BasePicker.propTypes);

      return result;
    }
  }, {
    key: 'defaultProps',
    get: function get() {
      var result = Object.assign({}, BasePicker.defaultProps);
      return result;
    }
  }]);

  function TimeSelect(props) {
    _classCallCheck(this, TimeSelect);

    // props, type, state
    return _possibleConstructorReturn(this, _BasePicker.call(this, props, 'timeselect', {}));
  }

  TimeSelect.prototype.isDateValid = function isDateValid(value) {
    return _BasePicker.prototype.isDateValid.call(this, value) && TimeSelectPanel.isValid(this.dateToStr(value), this.panelProps());
  };

  TimeSelect.prototype.panelProps = function panelProps(props) {
    var ps = props || this.props;
    var minTime = this.dateToStr(this.props.minTime);
    return _extends({}, ps, { minTime: minTime });
  };

  TimeSelect.prototype.pickerPanel = function pickerPanel(state, props) {
    var _this2 = this;

    var value = this.dateToStr(state.value);
    return React.createElement(TimeSelectPanel, _extends({}, this.panelProps(props), {
      value: value,
      onPicked: this.onPicked.bind(this),
      dateParser: function dateParser(str) {
        var r = _this2.parseDate(str);
        return r;
      }
    }));
  };

  return TimeSelect;
}(BasePicker);

export default TimeSelect;
