// element-react/dist/npm/es6/src/message/Toast.js
function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

import React from 'react';
import { Component, PropTypes, Transition, View } from '../../libs';
import icons from './assets';

var Toast = function (_Component) {
  _inherits(Toast, _Component);

  function Toast(props) {
    _classCallCheck(this, Toast);

    var _this = _possibleConstructorReturn(this, _Component.call(this, props));

    _this.state = {
      visible: false
    };
    return _this;
  }

  Toast.prototype.componentDidMount = function componentDidMount() {
    this.setState({
      visible: true
    });

    this.startTimer();
  };

  Toast.prototype.componentWillUnmount = function componentWillUnmount() {
    this.stopTimer();
  };

  Toast.prototype.onClose = function onClose() {
    var _this2 = this;

    this.stopTimer();

    this.setState({
      visible: false
    }, function () {
      _this2.props.willUnmount();
    });
  };

  Toast.prototype.startTimer = function startTimer() {
    var _this3 = this;

    if (this.props.duration > 0) {
      this.timeout = setTimeout(function () {
        _this3.onClose();
      }, this.props.duration);
    }
  };

  Toast.prototype.stopTimer = function stopTimer() {
    clearTimeout(this.timeout);
  };

  Toast.prototype.render = function render() {
    var _props = this.props,
        iconClass = _props.iconClass,
        customClass = _props.customClass;


    return React.createElement(
      Transition,
      { name: 'el-message-fade', duration: '300' },
      React.createElement(
        View,
        { key: this.state.visible, show: this.state.visible },
        React.createElement(
          'div',
          { className: this.classNames('el-message', customClass), onMouseEnter: this.stopTimer.bind(this), onMouseLeave: this.startTimer.bind(this) },
          !iconClass && React.createElement('img', { className: 'el-message__img', src: icons[this.props.type] }),
          React.createElement(
            'div',
            { className: this.classNames('el-message__group', { 'is-with-icon': iconClass }) },
            iconClass && React.createElement('i', { className: this.classNames('el-message__icon', iconClass) }),
            React.createElement(
              'p',
              null,
              this.props.message
            ),
            this.props.showClose && React.createElement('div', { className: 'el-message__closeBtn el-icon-close', onClick: this.onClose.bind(this) })
          )
        )
      )
    );
  };

  return Toast;
}(Component);

export default Toast;


Toast.propTypes = {
  type: PropTypes.oneOf(['success', 'warning', 'info', 'error']),
  message: PropTypes.string.isRequired,
  duration: PropTypes.number,
  showClose: PropTypes.bool,
  customClass: PropTypes.string,
  iconClass: PropTypes.string
};

Toast.defaultProps = {
  type: 'info',
  duration: 3000,
  showClose: false
};
