// element-react/dist/npm/es6/libs/editor/index.js
function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

import React, { Component } from 'react';
import PropTypes from 'prop-types';
import CodeMirror from 'codemirror';

import 'codemirror/mode/jsx/jsx';
import 'codemirror/keymap/sublime';
import 'codemirror/addon/comment/comment';

import 'codemirror/lib/codemirror.css';
import './style.scss';

var Editor = function (_Component) {
  _inherits(Editor, _Component);

  function Editor() {
    _classCallCheck(this, Editor);

    return _possibleConstructorReturn(this, _Component.apply(this, arguments));
  }

  Editor.prototype.componentDidMount = function componentDidMount() {
    var _props = this.props,
        onChange = _props.onChange,
        value = _props.value;


    this.cm = CodeMirror(this.editor, {
      mode: 'jsx',
      theme: 'react',
      keyMap: 'sublime',
      viewportMargin: Infinity,
      lineNumbers: false,
      dragDrop: false
    });

    this.cm.setValue(value);

    this.cm.on('changes', function (cm) {
      onChange && onChange(cm.getValue());
    });
  };

  Editor.prototype.render = function render() {
    var _this2 = this;

    return React.createElement('div', { className: 'editor', ref: function ref(_ref) {
        return _this2.editor = _ref;
      } });
  };

  return Editor;
}(Component);

export default Editor;


Editor.propTypes = {
  onChange: PropTypes.func,
  value: PropTypes.string
};
