// element-react/dist/npm/es6/src/upload/Cover.js
function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

import React from 'react';
import { Component, PropTypes } from '../../libs';

var Cover = function (_Component) {
  _inherits(Cover, _Component);

  function Cover(props) {
    _classCallCheck(this, Cover);

    var _this = _possibleConstructorReturn(this, _Component.call(this, props));

    _this.state = {
      dragOver: false
    };
    return _this;
  }

  Cover.prototype.handleDragover = function handleDragover(e) {
    e.preventDefault();
    this.setState({ dragOver: true });
  };

  Cover.prototype.handleDragleave = function handleDragleave(e) {
    e.preventDefault();
    this.setState({ dragOver: false });
  };

  Cover.prototype.onDrop = function onDrop(e) {
    e.preventDefault();
    this.setState({ dragOver: false });
    this.props.onFile(e.dataTransfer.files);
  };

  Cover.prototype.render = function render() {
    var _this2 = this;

    var dragOver = this.state.dragOver;

    return React.createElement(
      'div',
      {
        className: this.classNames({
          'el-upload-dragger': true,
          'is-dragover': dragOver
        }),
        onDrop: function onDrop(e) {
          return _this2.onDrop(e);
        },
        onDragOver: function onDragOver(e) {
          return _this2.handleDragover(e);
        },
        onDragLeave: function onDragLeave(e) {
          return _this2.handleDragleave(e);
        }
      },
      this.props.children
    );
  };

  return Cover;
}(Component);

Cover.defaultProps = {
  onFile: Function
};
export default Cover;


Cover.propTypes = {
  onFile: PropTypes.func
};
